phoenix-core/src/main/java/org/apache/phoenix/expression/util/regex/AbstractBasePattern.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.expression.util.regex;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.phoenix.schema.SortOrder;
+
+public abstract class AbstractBasePattern {
```

Comment:
I thinks AbstractBasePattern is necessary, for the following reasons:

1. Both j.u.regex and JONI has a pattern and match class.  AbstractBasePattern and AbstractBaseMatcher are abstraction of them. It is also quite easy to update existing code, simply replacing corresponding pattern and match to the abstract ones.

2. We might want to compile one pattern, but use this pattern to match multiple strings, and in this case, it would be easy to use.